examples_tests/20.Megatexture/main.cpp

```diff
@@ -76,359 +51,87 @@ layout (push_constant) uniform Block {
-		clippedLoD = highNotInLastFull ? (clippedLoD+1):clippedLoD;
-		levelInTail = highNotInLastFull ? levelInTail:(levelInTail+1);
-		loPhysCoord = vTextureGrad_helper(virtualUV,clippedLoD,levelInTail);
-	}
+#define _IRR_TEXTURE_SAMPLE_FUNCTIONS_DEFINED_
 
-	vec4 hiMip_retval;
-    vec4 loMip;
-#ifdef IRR_GL_EXT_nonuniform_qualifier
-    hiMip_retval = textureGrad(physPgTex[nonuniformEXT(formatID)],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
-    if (haveToDoTrilinear)
-        loMip = textureGrad(physPgTex[nonuniformEXT(formatID)],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
-#else
-    uint64_t outstandingSampleMask = ballotARB(true);
-    // maybe unroll a few times manually
-    while (outstandingSampleMask!=uint64_t(0u))
-    {
-		uvec2 tmp = unpackUint2x32(outstandingSampleMask);
-        uint subgroupFormatID = readInvocationARB(formatID,tmp[1]!=0u ? 32u:findLSB(tmp[0]));
-        bool canSample = subgroupFormatID==formatID; // do I need this? && (outstandingSampleMask&gl_SubGroupEqMaskARB)==gl_SubGroupEqMaskARB;
-        outstandingSampleMask ^= ballotARB(canSample);
-        if (canSample)
-        {
-            hiMip_retval = textureGrad(physPgTex[subgroupFormatID],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
-            if (haveToDoTrilinear)
-                loMip = textureGrad(physPgTex[subgroupFormatID],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
-        }
-    }
-#endif
-    if (haveToDoTrilinear)
-	    hiMip_retval = mix(hiMip_retval,loMip,LoD-float(LoD_high));
-    return hiMip_retval;
-}
+)";
+constexpr const char* GLSL_VT_FUNCTIONS =
+R"(
 
-uvec2 unpackWrapModes(in uvec2 texData)
+uint getPhysicalIDForLayer(in uint layer)
 {
-    return (texData>>uvec2(28u,30u)) & uvec2(0x03u);
+    return layer2pid.lut[layer];
 }
-uint unpackMaxMipInVT(in uvec2 texData)
-{
-    return (texData.y>>24)&0x0fu;
-}
-vec3 unpackVirtualUV(in uvec2 texData)
+
+uint getPgTabSzLog2(in uint _formatID)
 {
-	// assert that PAGE_SZ_LOG2<8 , or change the line to uvec3(texData.yy<<uvec2(PAGE_SZ_LOG2,PAGE_SZ_LOG2-8u),texData.y>>16u)
-    uvec3 unnormCoords = uvec3(texData.y<<PAGE_SZ_LOG2,texData.yy>>uvec2(8u-PAGE_SZ_LOG2,16u))&uvec3(uvec2(0xffu)<<PAGE_SZ_LOG2,0xffu);
-    return vec3(unnormCoords);
+    return precomputed.pgtab_sz_log2[_formatID];
 }
-vec2 unpackSize(in uvec2 texData)
+float getPhysPgTexSzRcp(in uint _formatID)
 {
-	return vec2(texData.x&0xffffu,texData.x>>16u);
+    return precomputed.phys_pg_tex_sz_rcp[_formatID];
 }
-#define WRAP_REPEAT 0u
-#define WRAP_CLAMP 1u
-#define WRAP_MIRROR 2u
-float wrapTexCoord(float tc, in uint mode)
+float getVTexSzRcp(in uint _formatID)
 {
-    switch (mode)
-    {
-    case WRAP_REPEAT: tc = fract(tc); break;
-    case WRAP_CLAMP:  tc = clamp(tc, 0.0, 1.0); break;
-    case WRAP_MIRROR: tc = 1.0 - abs(mod(tc,2.0)-1.0); break;
-    default: break;
-    }
-    return tc;
+    return precomputed.vtex_sz_rcp[_formatID];
 }
```

Comment:
Page size is constant, because page table is a 2D array that would mean that the virtual texture reciprocal size is identical always (constant/uniform not depending on formatID)